scvi/metrics/imputation.py

```diff
@@ -1,18 +1,118 @@
+import copy
+
+import matplotlib.pyplot as plt
 import numpy as np
 import torch
+from scipy.stats import kde
+
+from scvi.dataset import GeneExpressionDataset
+from scvi.dataset.data_loaders import DataLoaderWrapper
+
+
+def imputation(infer, name, rate=0.1, n_samples=1, n_epochs=1, corruption="uniform"):
+    corrupted_data = copy.deepcopy(infer.gene_dataset.X)
+
+    if corruption == "uniform":  # multiply the entry n with a Ber(0.9) random variable.
+        i, j = np.nonzero(corrupted_data)
+        ix = np.random.choice(range(len(i)), int(np.floor(rate * len(i))), replace=False)
+        i, j = i[ix], j[ix]
+        corrupted_data[i, j] *= np.random.binomial(n=np.ones(len(ix), dtype=np.int64), p=0.9)
+    elif corruption == "binomial":  # multiply the entry n with a Bin(n, 0.9) random variable.
+        i, j = (k.ravel() for k in np.indices(corrupted_data.shape))
+        ix = np.random.choice(range(len(i)), int(np.floor(rate * len(i))), replace=False)
+        i, j = i[ix], j[ix]
+        corrupted_data[i, j] = np.random.binomial(n=corrupted_data[i, j], p=0.2)
+
+    infer.gene_dataset = gene_dataset = GeneExpressionDataset(
+        *GeneExpressionDataset.get_attributes_from_matrix(
+            corrupted_data,
+            batch_indices=infer.gene_dataset.batch_indices,
+            labels=infer.gene_dataset.labels
+        )
+    )
+
+    original_data_loaders_loop = infer.data_loaders.loop
+    infer.data_loaders.loop = ['corrupted_%s' % s for s in infer.data_loaders.loop]
+    original_keys = list(infer.data_loaders.dict.keys())
+    for key in original_keys:
+        kwargs = copy.copy(infer.data_loaders.kwargs)
+        kwargs['collate_fn'] = gene_dataset.collate_fn
+        kwargs['sampler'] = copy.copy(infer.data_loaders[key].sampler)
+        infer.data_loaders['corrupted_%s' % key] = DataLoaderWrapper(gene_dataset, use_cuda=infer.use_cuda, **kwargs)
+
+    infer.train(n_epochs=n_epochs)
+    infer.data_loaders.loop = original_data_loaders_loop
+
+    original_list = []
+    imputed_list = []
+    batch_size = infer.data_loaders.kwargs["batch_size"] // n_samples
+    for tensors, corrupted_tensors in \
+        zip(infer.data_loaders[name].sequential(batch_size=batch_size),
```

Comment:
This is where I used the `.sequential` method and the SequentialSubsetSampler needed to iterate simultaneously on the same indices for both the uncorrupted and corrupted data.